### PR TITLE
New role: HOIST_ADMIN_READER

### PR DESCRIPTION
**Pull request by cnrudd**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [ ] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [ ] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR: https://github.com/xh/toolbox/pull/607
- [x] Created Hoist-Core branch / PR: https://github.com/xh/hoist-core/pull/271

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by PeteDarinzo** [comment]
Reviewed and reported JSON formatting issue

**Review by PeteDarinzo — APPROVED** [review]
Reviewed and reported JSON formatting issue

File: format/FormatMisc.js [hunk 1]
```diff
@@ -42,6 +42,21 @@ export function fmtSpan(v, {
     return asHtml ? fmtSpanHtml(v, opts) : fmtSpanElement(v, opts);
 }
 
+/**
+ * Format JSON string
+ *
+ * @param {string} str - string to be formatted into JSON
```
Comment (amcclain) [inline format/FormatMisc.js:48]: Ideally we would call this something like `prettyPrintJson()` but we have a strong convention of the leading `fmt` on these formatter function names.

We could clarify in the comment:
```
**
 * Pretty-print a JSON string, adding line breaks and indentation.
 *
 * @param {string} str - string to be pretty-printed, must be valid JSON.
 * ....
```

File: format/FormatMisc.js [hunk 2]
```diff
@@ -42,6 +42,21 @@ export function fmtSpan(v, {
...
+ * @param {Object} [opts] - an options object:
+ * @param {function|array} [opts.replacer] - method or array pattern for replacing/skipping nodes
+ * @param {string|number} [opts.space] - set indentation
+ */
+export function fmtJson(str, {
+    replacer = undefined,
+    space = 2
+} = {}) {
+    return JSON.stringify(JSON.parse(str), replacer, space);
```
Comment (amcclain) [inline format/FormatMisc.js:57]: Should this return `''` for a `null` input?  

I think that would be a bit more consistent with how other formatters work - I doubt we would often want the string `'null'` returned.

**Review by amcclain — COMMENTED** [review]
Looks great - huge amount of work to carry this through all the various components!

File: admin/differ/DifferDetail.js [hunk 3]
```diff
@@ -23,14 +23,16 @@ export const differDetail = hoistCmp.factory({
     render({model}) {
         if (!model.record) return null;
 
+        const readonly = !XH.getUser().isHoistAdmin;
```
[the remark this answers is not in the document]
Comment (cnrudd) [reply admin/differ/DifferDetail.js]: Great idea!  Done, with bonus cleanups.

File: format/FormatMisc.js (hunk 1 above)
Comment (cnrudd) [reply format/FormatMisc.js]: Agreed.  jsDocs fixed.

File: format/FormatMisc.js (hunk 2 above)
Comment (cnrudd) [reply format/FormatMisc.js]: Agreed.  Done.

**Pull request merged**